File: server/Models/FinanceModels.cs
namespace PersonalFinances.Api.Models;

public enum TransactionType
{
    Income,
    Expense
}

public enum TagSelectionMode
{
    Single,
    Multiple
}

public sealed class FinanceTransaction
{
    public int Id { get; set; }
    public TransactionType Type { get; set; }
    public decimal Amount { get; set; }
    public DateOnly Date { get; set; }
    public string Description { get; set; } = string.Empty;
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public ICollection<TransactionTag> TransactionTags { get; set; } = [];
}

public sealed class TagGroup
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public TagSelectionMode SelectionMode { get; set; }
    public bool IsRequired { get; set; }
    public ICollection<Tag> Tags { get; set; } = [];
}

public sealed class Tag
{
    public int Id { get; set; }
    public int TagGroupId { get; set; }
    public TagGroup TagGroup { get; set; } = null!;
    public string Name { get; set; } = string.Empty;
    public string Color { get; set; } = "#64748b";
    public int? ParentTagId { get; set; }
    public Tag? ParentTag { get; set; }
    public ICollection<Tag> Children { get; set; } = [];
    public ICollection<TransactionTag> TransactionTags { get; set; } = [];
    public ICollection<BudgetItemTag> BudgetItemTags { get; set; } = [];
    public ICollection<FixedExpenseTag> FixedExpenseTags { get; set; } = [];
}

public sealed class TransactionTag
{
    public int TransactionId { get; set; }
    public FinanceTransaction Transaction { get; set; } = null!;
    public int TagId { get; set; }
    public Tag Tag { get; set; } = null!;
}

public sealed class Budget
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public DateOnly From { get; set; }
    public DateOnly To { get; set; }
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public ICollection<BudgetItem> Items { get; set; } = [];
}

public sealed class BudgetItem
{
    public int Id { get; set; }
    public int BudgetId { get; set; }
    public Budget Budget { get; set; } = null!;
    public string Description { get; set; } = string.Empty;
    public TransactionType Type { get; set; } = TransactionType.Expense;
    public decimal ExpectedAmount { get; set; }
    public ICollection<BudgetItemTag> BudgetItemTags { get; set; } = [];
}

public sealed class BudgetItemTag
{
    public int BudgetItemId { get; set; }
    public BudgetItem BudgetItem { get; set; } = null!;
    public int TagId { get; set; }
    public Tag Tag { get; set; } = null!;
}

public sealed class FixedExpense
{
    public int Id { get; set; }
    public TransactionType Type { get; set; } = TransactionType.Expense;
    public string Description { get; set; } = string.Empty;
    public decimal Amount { get; set; }
    public int Month { get; set; }
    public int? ForecastWeek { get; set; }
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public ICollection<FixedExpenseTag> FixedExpenseTags { get; set; } = [];
}

public sealed class FixedExpenseTag
{
    public int FixedExpenseId { get; set; }
    public FixedExpense FixedExpense { get; set; } = null!;
    public int TagId { get; set; }
    public Tag Tag { get; set; } = null!;
}

public sealed class DriveSettings
{
    public int Id { get; set; } = 1;
    public string? ClientId { get; set; }
    public string? ClientSecret { get; set; }
    public string? RefreshToken { get; set; }
    public string? ConnectedAccountEmail { get; set; }
    public string? FolderId { get; set; }
    public bool AutoUpload { get; set; } = true;
}

public sealed class MonthlyFixedExpense
{
    public int Id { get; set; }
    public string Description { get; set; } = string.Empty;
    public decimal Amount { get; set; }
    public int Week { get; set; }
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
}

File: client/src/app/finance.models.ts
export type TransactionType = 'income' | 'expense';
export type SummaryPeriod = 'week' | 'month' | 'year';

export interface Tag {
  id: number;
  name: string;
  color: string;
  parentTagId: number | null;
  linkedSavingsAccountId: number | null;
}

export interface TagGroup {
  id: number;
  name: string;
  selectionMode: 'single' | 'multiple';
  isRequired: boolean;
  tags: Tag[];
}

export interface TransactionTag {
  id: number;
  groupId: number;
  groupName: string;
  name: string;
  color: string;
}

export interface FinanceTransaction {
  id: number;
  type: TransactionType;
  amount: number;
  date: string;
  description: string;
  tags: TransactionTag[];
}

export interface TagTotal {
  tagId: number;
  tagName: string;
  groupName: string;
  color: string;
  income: number;
  expense: number;
  balance: number;
}

export interface FinanceSummary {
  from: string;
  to: string;
  income: number;
  expense: number;
  balance: number;
  tagTotals: TagTotal[];
}

export interface WeeklyForecastWeek {
  weekNumber: number;
  from: string;
  to: string;
  isCurrent: boolean;
  isEstimate: boolean;
  forecastIncome: number;
  forecastExpense: number;
  actualIncome: number;
  actualExpense: number;
  remainingBalance: number;
  tagTotals: TagTotal[];
}

export interface WeeklyForecast {
  monthFrom: string;
  monthTo: string;
  monthIncome: number;
  incomeSource: 'budget' | 'actual';
  remainingWeeklyBudget: number;
  remainingWeeksCount: number;
  weeks: WeeklyForecastWeek[];
}

export interface MonthlyFixedExpense {
  id: number;
  description: string;
  amount: number;
  week: number;
}

export interface CreateMonthlyFixedExpense {
  description: string;
  amount: number;
  week: number;
}

export interface CreateTransaction {
  type: TransactionType;
  amount: number;
  date: string;
  description: string;
  tagIds: number[];
}

export interface CreateTag {
  tagGroupId: number;
  name: string;
  color: string;
  parentTagId: number | null;
}

export interface BudgetItem {
  id: number;
  type: TransactionType;
  description: string;
  expectedAmount: number;
  tags: TransactionTag[];
}

export interface Budget {
  id: number;
  name: string;
  from: string;
  to: string;
  expectedIncome: number;
  expectedExpense: number;
  expectedBalance: number;
  items: BudgetItem[];
  comparisons: BudgetComparison[];
}

export interface BudgetComparison {
  key: string;
  typeName: string;
  subtypeName: string;
  expectedExpense: number;
  actualExpense: number;
  difference: number;
  usagePercentage: number;
  isExceeded: boolean;
}

export interface CreateBudget {
  name: string;
  from: string;
  to: string;
}

export interface CreateBudgetItem {
  type: TransactionType;
  description: string;
  expectedAmount: number;
  tagIds: number[];
}

export interface FixedExpense {
  id: number;
  type: TransactionType;
  description: string;
  amount: number;
  month: number;
  forecastWeek: number | null;
  tags: TransactionTag[];
}

export interface CreateFixedExpense {
  type: TransactionType;
  description: string;
  amount: number;
  month: number;
  tagIds: number[];
}

export interface SaveFixedExpenseForecastWeek {
  week: number | null;
}

export type SavingsMovementType = 'deposit' | 'withdrawal';

export interface SavingsAccount {
  id: number;
  name: string;
  color: string;
  balance: number;
  totalDeposits: number;
  totalWithdrawals: number;
}

export interface CreateSavingsAccount {
  name: string;
  color: string;
}

export type SavingsMovementSource = 'manual' | 'transaction';

export interface SavingsMovement {
  id: number;
  savingsAccountId: number;
  savingsAccountName: string;
  type: SavingsMovementType;
  amount: number;
  date: string;
  description: string;
  source: SavingsMovementSource;
}

export interface CreateSavingsMovement {
  savingsAccountId: number;
  type: SavingsMovementType;
  amount: number;
  date: string;
  description: string;
}

export interface SaveTagSavingsLink {
  savingsAccountId: number | null;
}

export interface BackupResult {
  fileName: string;
  relativePath: string;
  sizeBytes: number;
  createdAt: string;
}
